src/lib/protocols/quic.c

```diff
@@ -1002,41 +1002,68 @@ static uint8_t *decrypt_initial_packet(struct ndpi_detection_module_struct *ndpi
   return NULL;
 }
 
+static void update_reasm_buf_bitmap(u_int8_t *buffer_bitmap,
+                                const u_int32_t buffer_bitmap_size,
+                                const u_int32_t recv_pos,
+                                const u_int32_t recv_len)
+{
+  if (!recv_len || !buffer_bitmap_size || recv_pos + recv_len > buffer_bitmap_size * 8)
+    return;
+  const u_int32_t start_byte = recv_pos / 8;
+  const u_int32_t end_byte = (recv_pos + recv_len - 1) / 8;
+  const u_int32_t start_bit = recv_pos % 8;
+  const u_int32_t end_bit = (start_bit + recv_len - 1) % 8;
+  if (start_byte == end_byte)
+    buffer_bitmap[start_byte] |= (((1U << recv_len) - 1U) << start_bit); // fill from bit 'start_bit' until bit 'end_bit', both inclusive
+  else{
+    for (u_int32_t i = start_byte + 1; i <= end_byte - 1; i++)
+      buffer_bitmap[i] = 0xff; // completely received byte
+    buffer_bitmap[start_byte] |= ~((1U << start_bit) - 1U); // fill from bit 'start_bit' until bit 7, both inclusive
+    buffer_bitmap[end_byte] |= (1U << (end_bit + 1U)) - 1U; // fill from bit 0 until bit 'end_bit', both inclusive
+  }
+}
+
+static int is_reasm_buf_complete(const u_int8_t *buffer_bitmap,
+                                 const u_int32_t buffer_len)
+{
+  for(u_int32_t i = 0; i < buffer_len / 8; i++)
+    if (buffer_bitmap[i] != 0xff)
+      return 0;
+  return 1;
+}
 
 static int __reassemble(struct ndpi_flow_struct *flow, const u_int8_t *frag,
                         uint64_t frag_len, uint64_t frag_offset,
                         const u_int8_t **buf, u_int64_t *buf_len)
 {
-  const uint64_t max_quic_reasm_buffer_len = 4096; /* Let's say a couple of full-MTU packets... */
+  const uint64_t max_quic_reasm_buffer_len = 4096; /* Let's say a couple of full-MTU packets... Must be multiple of 8*/
+  const uint64_t quic_reasm_buffer_bitmap_len = max_quic_reasm_buffer_len / 8;
   const uint64_t last_pos = frag_offset + frag_len;
-  /* 
-     Working: in-order and out-of order, non overlapping, fragments
-     Not supported: retransmissions and (partial) overlapping
-  */
 
   if(!flow->l4.udp.quic_reasm_buf) {
     flow->l4.udp.quic_reasm_buf = (uint8_t *)ndpi_malloc(max_quic_reasm_buffer_len);
-    if(!flow->l4.udp.quic_reasm_buf)
+    flow->l4.udp.quic_reasm_buf_bitmap = (uint8_t *)ndpi_malloc(quic_reasm_buffer_bitmap_len);
```

Comment:
What about `ndpi_calloc`?

Reply:
I changed both mallocs to ndpi_calloc.

Reply:
Please don't. 
`ndpi_calloc()` with `quic_reasm_buf_bitmap` is fine because you were going to zero it anyway.
But I strongly prefer (if it is possible) to NOT initialize `udp.quic_reasm_buf` and let Valgrind/MASAN to warning us if we do some invalid read: we should read the buffer only if it is valid; if you "calloc" it we loose this "feature".

Reply:
I got it. I will change it back and check Valgrind results.